WIP Initial PowerShelll kernel prototype

TylerLeonhardt opened the pull request:
This handles the 3 basic messages:
* HandleSubmitCode
* HandleRequestCompletion
* HandleCancelCurrentCommand (but is just a no-op at the moment)

Things pending in this PR:
* Handling cancel
* tests 😃 

Example [PowerShell notebook](https://gist.github.com/TylerLeonhardt/ff9a20fb53313bf31e6c0107b0fec4e3) that came from this very change.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellExtensions.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace Microsoft.DotNet.Interactive.PowerShell
+{
+    using System.Management.Automation;
```
Nit: You can move the using outside the namespace declaration.

TylerLeonhardt replied:
Unfortunately not for that one. That's because the type that I need to use is called `PowerShell` and the namespace is called `Microsoft.DotNet.Interactive.PowerShell` so they conflict.

TylerLeonhardt commented:
The "order matters" issue can be fixed by using a similar solution to what we did with Azure Functions. Dropping this file for future reference
https://github.com/Azure/azure-functions-powershell-worker/blob/dev/src/Public/Commands/TracePipelineObjectCommand.cs

SteveL-MSFT commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,347 @@
+        static PowerShellKernel()
+        {
+            Formatter<WarningRecord>.Register((record, writer) => {
+                PocketView view = PocketViewTags.pre($"WARN: {record.Message}");
```
pwsh writes this out as `WARNING:`

SteveL-MSFT commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,347 @@
+            }, HtmlFormatter.MimeType);
+
+            Formatter<WarningRecord>.Register((record, writer) => {
+                writer.WriteLine($"WARN: {record.Message}");
```
Same as above

SteveL-MSFT commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,347 @@
+                        {
+                            results = ps
+                                .AddScript(code)
+                                .AddCommand("Out-String")
```
Should this be fully qualified?

SteveL-MSFT commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,347 @@
+            IKernelCommand command)
+        {
+            var formattedValues = new List<FormattedValue>
+                        {
```
Indentation here is off

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,347 @@
+            }
+            else
+            {
+                context.Publish(new CommandFailed(null, submitCode, "Command cancelled"));
```
It's preferable here to do `context.Fail`. Directly publishing these commands might be something we do away with.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/StreamHandler.cs:
```diff
@@ -0,0 +1,180 @@
+            // InformationRecord
+            Formatter<InformationRecord>.Register((record, writer) => {
+                string prefix = (record.Tags.Count == 1 && record.Tags[0] == "__PipelineObject__") ? "OUTPUT:" : "INFORMATION:";
+                PocketView view = PocketViewTags.pre($"{prefix} {record.MessageData}");
```
You can add `using static Microsoft.DotNet.Interactive.Formatting.PocketViewTags` at the top of the file so you don't have to fully qualify calls to `pre`, etc.

TylerLeonhardt commented:
Haven't addressed feedback yet but have addressed the "order matters" problem. See the example Notebook (in GitHub gists) to see the results of that work. It's using the same concept that Azure Functions to maintain order (a `Trace-PipelineObject` cmdlet).

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/Commands/TracePipelineObjectCommand.cs:
```diff
@@ -0,0 +1,69 @@
+        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true)]
+        public object InputObject { get; set; }
+
+        private static PowerShell s_pwsh;
```
Nit: We don't follow the `s`-prefix convention.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,311 @@
+            string psModulePath = Environment.GetEnvironmentVariable("PSModulePath");
+            Environment.SetEnvironmentVariable("PSModulePath",
+                psModulePath + Path.PathSeparator +
+                System.IO.Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) + "/Modules");
```
You probably want to remove "/" from "/Modules".

Using `Path.Combine` will let you avoid having to specify the path separators.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,311 @@
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            context.Publish(
```
We added a `KernelInvocationContext.DisplayAsync` method that might simplify this for you.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,311 @@
+            IKernelCommand command)
+        {
+            var streamHandler = new StreamHandler(context, command);
+            powerShell.Streams.Debug.DataAdding += streamHandler.DebugDataAdding;
```
Are you unhooking these events anywhere?

Please make sure to dispose anything that needs disposing.

TylerLeonhardt replied:
These events are hooked up to the PowerShell object that gets disposed via the `using(...)`. I could unhook them if you'd like but it shouldn't be needed.

jonsequitur replied:
It should usually be fine. We just spent too much time tracking down a memory leak resulting from a dangling object reference, which could conceivably happen here.

TylerLeonhardt replied:
I can also clean them up 👍

TylerLeonhardt commented:
Closing in favor of https://github.com/dotnet/interactive/pull/9
